universal-application-tool-0.0.1/app/controllers/admin/QuestionController.java

```diff
@@ -90,41 +90,43 @@ public QuestionController(
               try {
                 return ok(editView.renderViewQuestionForm(questionDefinition));
               } catch (InvalidQuestionTypeException e) {
-                return badRequest(e.toString());
+                return badRequest(
+                    invalidQuestionTypeMessage(questionDefinition.getQuestionType().toString()));
               }
             },
             httpExecutionContext.current());
   }
 
   @Secure(authorizers = Authorizers.Labels.UAT_ADMIN)
   public Result newOne(Request request, String type) {
+    QuestionType questionType;
+    try {
+      questionType = QuestionType.of(type);
+    } catch (InvalidQuestionTypeException e) {
+      return badRequest(invalidQuestionTypeMessage(type));
+    }
+
+    ImmutableList<RepeaterQuestionDefinition> repeaterQuestionDefinitions =
+        service
+            .getReadOnlyQuestionService()
+            .toCompletableFuture()
+            .join()
+            .getUpToDateRepeaterQuestions();
+
     try {
-      QuestionType questionType = QuestionType.valueOf(type.toUpperCase());
-      ImmutableList<RepeaterQuestionDefinition> repeaterQuestionDefinitions =
-          service
-              .getReadOnlyQuestionService()
-              .toCompletableFuture()
-              .join()
-              .getUpToDateRepeaterQuestions();
       return ok(editView.renderNewQuestionForm(request, questionType, repeaterQuestionDefinitions));
     } catch (UnsupportedQuestionTypeException e) {
       return badRequest(e.getMessage());
-    } catch (IllegalArgumentException e) {
```

Comment:
question: what is throwing the `IllegalArgumentException`? Is it something in the view that still needs to be caught?

Reply:
Good question. I changed a line above from `QuestionType.valueOf` (which can throw an `IllegalArgumentException`), to using `QuestionType.of`, which is a method on `QuestionType` that does some nicer handling and potentially throws an `InvalidQuestionTypeException`. So this should be fine.